library/public/types.h

```diff
@@ -82,6 +74,145 @@ struct mesh_t
    */
   F3D_EXPORT std::pair<bool, std::string> isValid() const;
 };
+
+/**
+ * Describe a 3D vector.
+ */
+struct F3D_EXPORT vector3_t
+{
+  vector3_t() = default;
+  vector3_t(double x, double y, double z)
+    : Value{ x, y, z }
+  {
+  }
+  vector3_t(const std::vector<double>& vec)
+  {
+    if (vec.size() != 3)
+    {
+      throw std::runtime_error("cannot create a vector3_t");
+    }
+    Value[0] = vec[0];
+    Value[1] = vec[1];
+    Value[2] = vec[2];
+  }
+  vector3_t(const std::array<double, 3>& arr)
+  {
+    Value[0] = arr[0];
+    Value[1] = arr[1];
+    Value[2] = arr[2];
+  }
+  vector3_t(const double* ptr)
+  {
+    Value[0] = ptr[0];
+    Value[1] = ptr[1];
+    Value[2] = ptr[2];
+  }
+  vector3_t(std::initializer_list<double> l)
+  {
+    if (l.size() != 3)
+    {
+      throw std::runtime_error("cannot create a vector3_t");
+    }
+    std::copy(l.begin(), l.end(), std::begin(Value));
+  }
+
+  double* data()
+  {
+    return Value.data();
+  }
+  const double* data() const
+  {
+    return Value.data();
+  }
+
+  double& operator[](int idx)
+  {
+    return Value[idx];
+  }
+  double operator[](int idx) const
+  {
+    return Value[idx];
+  }
+  operator std::vector<double>() const
+  {
+    return { Value[0], Value[1], Value[2] };
+  }
+  operator std::array<double, 3>() const
+  {
+    return { Value[0], Value[1], Value[2] };
+  }
+  bool operator==(const vector3_t& vec) const
+  {
+    return Value[0] == vec.Value[0] && Value[1] == vec.Value[1] && Value[2] == vec.Value[2];
+  }
+  bool operator!=(const vector3_t& vec) const
+  {
+    return !(*this == vec);
+  }
+
+  auto begin()
+  {
+    return Value.begin();
+  }
+  auto begin() const
```

Comment:
I think you can remove `begin() const` and `end() const`. They are uncovered and I don't think they make sense anyway.